ext/src/http/client/curl/http_operation_curl.cc

```diff
@@ -0,0 +1,685 @@
+      curl_resource_.headers_chunk =
+          curl_slist_append(curl_resource_.headers_chunk, header.c_str());
+    }
+  }
+
+  DispatchEvent(opentelemetry::ext::http::client::SessionState::Created);
+}
+
+HttpOperation::~HttpOperation()
+{
+  // Given the request has not been aborted we should wait for completion here
+  // This guarantees the lifetime of this request.
+  switch (GetSessionState())
+  {
+    case opentelemetry::ext::http::client::SessionState::Connecting:
+    case opentelemetry::ext::http::client::SessionState::Connected:
+    case opentelemetry::ext::http::client::SessionState::Sending: {
+      if (async_data_ && async_data_->result_future.valid())
+      {
+        if (async_data_->callback_thread != std::this_thread::get_id())
+        {
+          async_data_->result_future.wait();
+          last_curl_result_ = async_data_->result_future.get();
+        }
+      }
+      break;
+    }
+    default:
+      break;
+  }
+
+  Cleanup();
+}
+
+void HttpOperation::Finish()
+{
+  if (is_finished_.exchange(true, std::memory_order_acq_rel))
+  {
+    return;
+  }
+
+  if (async_data_ && async_data_->result_future.valid())
+  {
+    // We should not wait in callback from Cleanup()
+    if (async_data_->callback_thread != std::this_thread::get_id())
+    {
+      async_data_->result_future.wait();
+      last_curl_result_ = async_data_->result_future.get();
+    }
+  }
+}
+
+void HttpOperation::Cleanup()
+{
+  if (is_cleaned_.exchange(true, std::memory_order_acq_rel))
+  {
+    return;
+  }
+
+  switch (GetSessionState())
+  {
+    case opentelemetry::ext::http::client::SessionState::Created:
+    case opentelemetry::ext::http::client::SessionState::Connecting:
+    case opentelemetry::ext::http::client::SessionState::Connected:
+    case opentelemetry::ext::http::client::SessionState::Sending: {
+      DispatchEvent(opentelemetry::ext::http::client::SessionState::Cancelled,
+                    curl_easy_strerror(last_curl_result_));
+      break;
+    }
+    default:
+      break;
+  }
+
+  std::function<void(HttpOperation &)> callback;
+
+  // Only cleanup async once even in recursive calls
+  if (async_data_)
+  {
+    // Just reset and move easy_handle to owner if in async mode
+    if (async_data_->session != nullptr)
+    {
+      auto session         = async_data_->session;
+      async_data_->session = nullptr;
+
+      if (curl_resource_.easy_handle != nullptr)
+      {
+        curl_easy_setopt(curl_resource_.easy_handle, CURLOPT_PRIVATE, NULL);
+        curl_easy_reset(curl_resource_.easy_handle);
+      }
+      session->GetHttpClient().ScheduleRemoveSession(session->GetSessionId(),
+                                                     std::move(curl_resource_));
+    }
+
+    callback.swap(async_data_->callback);
+    if (callback)
+    {
+      async_data_->callback_thread == std::this_thread::get_id();
+      callback(*this);
+      async_data_->callback_thread == std::thread::id();
```

Comment:
I am not clear why are we assigning a new thread id here?

Reply:
It's not necessary, but I think it's more clear to reset `callback_thread` after callback is finished.